### PR TITLE
Feature: add ibis meta data routers

### PR DESCRIPTION
**Pull request by onlyjackfrost**
add ibis meta data routers

### Request body
for postgres
```JSON
{
    "connectionInfo": {
        "host": "localhost",
        "port": "6432",
        "user": "postgres",
        "database": "postgres",
        "password": "postgres"
    }
}
```
for bigquery
```JSON
{
    "connectionInfo": {
        "project_id": "wrenai",
        "dataset_id": "wrenai.ecommerce",
        "credentials": "your base64 encoded JSON string of your credential file"
    }
}
```

### Response 
- name: unique table name (might contain schema name, depends on are you listing table across schema or not )
- columns
    - name: column name in the datasource
    - type: column data type
    - notNull: boolean, nullable or not
    - description: column description(comment) if any
    - properties: column properties if any
- description: table description if any
- properties
    - schema: schema name to build tableReference
    - catalog: catalog name to build tableReference
    - table: table name to build tableReference
- primaryKey: the column name which is bind with primary constraint
```JSON
{
   [
        {
            "name": "public.nation",
            "columns": [
                {
                    "name": "n_nationkey",
                    "type": "INTEGER",
                    "notNull": true,
                    "description": "",
                    "properties": {}
                }
            ],
            "description": "",
            "properties": {
                "schema": "public",
                "catalog": "postgres",
                "table": "nation"
            },
            "primaryKey": ""
        },
  ]
}
```

### Request body
Same as above

### Response 
- constraints
    - constraintName: unique constraint name  `({constraint_table}_{constraint_column}_{constrainted_table}_{constrainted_column})`
    - constraintType: “FOREIGN KEY”
    - constraintTable
    - constraintColumn
    - constraintedTable
    - constraintedColumn
```JSON
{
    [
        {
            "constraintName": "composite_pk_y_composite_fk_z",
            "constraintType": "FOREIGN KEY",
            "constraintTable": "composite_pk",
            "constraintColumn": "y",
            "constraintedTable": "composite_fk",
            "constraintedColumn": "z"
        }
    ]
}
```

### CODE REVIEW

**Comment by goldmedal** [comment]
Could you add some tests for them? It would make it more stable. Furthermore, could you add a description for the new API?

**Comment by grieve54706** [comment]
Plz rebase the main branch and use Ruff to format codes. Follow #601

File: ibis-server/app/model/metadata.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                    "description": "",
+                    "columns": [],
+                    "properties": {
+                        "schema": row["table_schema"],
+                        "catalog": row["table_catalog"],
+                        "table": row["table_name"],
+                    },
+                    "primaryKey": "",
+                }
+                unique_tables[schema_table] = CompactTable(**table)
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:92]: How about directly using class new?
```python
unique_tables[schema_table] = CompactTable(
    name=schema_table,
    properties=CompactTableProperties(
        schema=row["table_schema"],
        catalog=row["table_catalog"],
        table=row["table_name"],
    )
)
```
`CompactColumn` too.

File: ibis-server/app/model/metadata_dto.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    properties: Optional[Dict[str, Any]] = None
+
+
+class CompactTableProperties(BaseModel):
+    schema: Optional[str]
+    catalog: Optional[str]
+    table: Optional[str]  # only table name without schema or catalog
+
+
+class CompactTable(BaseModel):
```
Comment (grieve54706) [inline ibis-server/app/model/metadata_dto.py:94]: You can just name `Table` in the metadata_dto.py. Let's drop the redundant `Compact`

File: ibis-server/app/model/metadata.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+                "type": row["data_type"],
+                "notNull": row["is_nullable"].lower() == "no",
+                "description": "",
+                "properties": {},
+            }
+            column = CompactColumn(**column)
+            unique_tables[schema_table].columns.append(column)
+
+        compact_tables: list[CompactTable] = list(unique_tables.values())
+        return compact_tables
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:105]: You can just `return list(unique_tables.values())`.

File: ibis-server/app/model/metadata.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        if self == Metadata.postgres:
+            return self.get_postgres_table_constraints(connection_info)
+        if self == Metadata.bigquery:
+            return self.get_bigquery_table_constraints(connection_info)
+        raise NotImplementedError(f"Unsupported data source: {self}")
+
+    @staticmethod
+    def get_postgres_table_list_sql(
+        connection_info: PostgresConnectionUrl | PostgresConnectionInfo,
+    ):
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:35]: Add return type may be clear.

File: ibis-server/app/model/metadata.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+            JOIN {dataset_id}.INFORMATION_SCHEMA.COLUMN_FIELD_PATHS cf 
+                ON cf.table_name = c.table_name 
+                AND cf.column_name = c.column_name
+            LEFT JOIN {dataset_id}.INFORMATION_SCHEMA.TABLE_OPTIONS table_options
+                ON c.table_name = table_options.table_name
+            WHERE
+                cf.column_name = cf.field_path
+                AND NOT REGEXP_CONTAINS(cf.data_type, r'^(STRUCT|ARRAY<STRUCT)')
+            """
+        print(sql)
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:142]: Plz do not use `print`. Let's use `logger` instead of it.

File: ibis-server/app/model/metadata.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                information_schema.tables t
+            JOIN
+                information_schema.columns c ON t.table_schema = c.table_schema AND t.table_name = c.table_name
+            WHERE
+                t.table_type in ('BASE TABLE', 'VIEW')
+                and t.table_schema not in ('information_schema', 'pg_catalog')
+            """
+        res = to_json(
+            DataSource.postgres.get_connection(connection_info)
+            .sql(sql, dialect="trino")
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:55]: The dialect is for the sql if not the data source sql. You can use `dialect` to let ibis know your sql is what dialect to convert the sql to data source sql.
Here, you clearly use Postgres SQL. You don't provide dialect to ibis.

File: ibis-server/app/model/metadata.py [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+            LEFT JOIN {dataset_id}.INFORMATION_SCHEMA.TABLE_OPTIONS table_options
+                ON c.table_name = table_options.table_name
+            WHERE
+                cf.column_name = cf.field_path
+                AND NOT REGEXP_CONTAINS(cf.data_type, r'^(STRUCT|ARRAY<STRUCT)')
+            """
+        print(sql)
+        res = to_json(
+            DataSource.bigquery.get_connection(connection_info)
+            .sql(sql, dialect="bigquery")
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:145]: SQL is BigQuery SQL. We don't provide dialect for ibis converting sql from the specific dialect.

File: ibis-server/app/model/metadata.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+                    "foreign_table_schema": x[3],
+                    "foreign_table_name": x[4],
+                    "foreign_column_name": x[5],
+                }
+            )(row)
+            for row in res["data"]
+        ]
+        constraints = []
+        for row in res:
+            constraint_table = formatPostgresCompactTableName(
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:251]: Plz use snake case instead of camel case.

File: ibis-server/app/model/metadata.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+            for row in res["data"]
+        ]
+        constraints = [Constraint(**row) for row in response]
+        return constraints
+
+
+def to_json(df):
+    json_obj = loads(df.to_json(orient='split'))
+    del json_obj['index']
+    json_obj['dtypes'] = df.dtypes.apply(lambda x: x.name).to_dict()
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:314]: You don't use `dtypes` here. Plz remove unused code.

File: ibis-server/app/model/metadata.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+                    "ordinal_position": x[6],
+                }
+            )(row)
+            for row in res["data"]
+        ]
+
+        unique_tables = {}
+        for row in response:
+            # generate unique table name
+            schema_table = f"{row['table_schema']}.{row['table_name']}"
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:78]: How about using the same func as L#247

File: ibis-server/app/model/metadata.py [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+            row = {
+                "constraintName": f"{table_name}_{column_name}_{foreign_table_name}_{foreign_column_name}",
+                "constraintTable": constraint_table,
+                "constraintColumn": column_name,
+                "constraintedTable": constrainted_table,
+                "constraintedColumn": foreign_column_name,
+                "constraintType": ConstraintType.FOREIGN_KEY,
+            }
+            constraints.append(Constraint(**row))
+        return constraints
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:270]: How about integrating them?
```python
return [Constraint(
            constraintName=f"{row['table_name']}_{row['column_name']}_{row['foreign_table_name']}_{row['foreign_column_name']}",
            constraintTable=formatPostgresCompactTableName(row["table_schema"], row["table_name"]),
            constraintColumn=row["column_name"],
            constraintedTable=formatPostgresCompactTableName(row["foreign_table_schema"], row["foreign_table_name"]),
            constraintedColumn=row["foreign_column_name"],
            constraintType=ConstraintType.FOREIGN_KEY,
        ) for row in res]
```

File: ibis-server/app/model/metadata_dto.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    UNIQUE = "UNIQUE"
+
+
+class Constraint(BaseModel):
+    constraintName: str
+    constraintType: ConstraintType
+    constraintTable: str
+    constraintColumn: str
+    constraintedTable: str
+    constraintedColumn: str
```
Comment (grieve54706) [inline ibis-server/app/model/metadata_dto.py:114]: Let's remove the redundant prefix `constraint`.

File: ibis-server/app/model/metadata.py [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+                    "constraintColumn": x[2],
+                    "constraintedTable": x[3],
+                    "constraintedColumn": x[4],
+                    "constraintType": ConstraintType.FOREIGN_KEY,
+                }
+            )(row)
+            for row in res["data"]
+        ]
+        constraints = [Constraint(**row) for row in response]
+        return constraints
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:308]:
```python
return [
    Constraint(
        constraintName=row[0],
        constraintTable=row[1],
        constraintColumn=row[2],
        constraintedTable=row[3],
        constraintedColumn=row[4],
        constraintType=ConstraintType.FOREIGN_KEY,
    )
    for row in res["data"]
]
```

File: ibis-server/app/model/metadata.py [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+
+
+def formatPostgresCompactTableName(schema, table):
+    return f"{schema}.{table}"
+
+
+def transform_postgres_column_type(data_type):
+    # lower case the data_type
+    data_type = data_type.lower()
+    print(f"=== data_type: {data_type}")
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:325]: Don't use print

File: ibis-server/app/model/metadata.py [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+        "timestamp without time zone": WrenEngineColumnType.TIMESTAMP,
+        "timestamp with time zone": WrenEngineColumnType.TIMESTAMPTZ,
+        "date": WrenEngineColumnType.DATE,
+        "interval": WrenEngineColumnType.INTERVAL,
+        "json": WrenEngineColumnType.JSON,
+        "bytea": WrenEngineColumnType.BYTEA,
+        "uuid": WrenEngineColumnType.UUID,
+        "inet": WrenEngineColumnType.INET,
+        "oid": WrenEngineColumnType.OID,
+    }
```
Comment (grieve54706) [inline ibis-server/app/model/metadata.py:355]: Use enum mapping.
```python
class ColumnType(Enum):
    TEXT = ("TEXT", "text")

    def __init__(self, wtype, ptype):
        self.wtype = wtype
        self.ptype = ptype

    @property
    def wtype(self):
        return self.wtype

    @property
    def ptype(self):
        return self.ptype
```

File: ibis-server/app/model/metadata_dto.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+from pydantic import BaseModel, Field
+from typing import List, Optional, Dict, Any, Union
+
+
+class MetadataDTO(BaseModel):
+    connection_info: Union[
+        PostgresConnectionUrl | PostgresConnectionInfo,
+        BigQueryConnectionInfo,
+        SnowflakeConnectionInfo,
+    ] = Field(alias="connectionInfo")
```
Comment (grieve54706) [inline ibis-server/app/model/metadata_dto.py:17]: You could just use `connection_info: ConnectionInfo = Field(alias="connectionInfo")`

File: ibis-server/app/model/metadata_dto.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    BYTEA = "BYTEA"
+
+    # UUID Type
+    UUID = "UUID"
+
+    # Network Address Types
+    INET = "INET"
+
+    # Unknown Type
+    UNKNOWN = "UNKNOWN"
```
Comment (grieve54706) [inline ibis-server/app/model/metadata_dto.py:77]: Plz remove too many space lines.

File: ibis-server/app/routers/ibis/bigquery.py [hunk 18]
```diff
@@ -21,3 +23,17 @@ def query(dto: BigQueryDTO, dry_run: Annotated[bool, Query(alias="dryRun")] = Fa
...
+def get_bigquery_table_list(dto: MetadataDTO) -> dict:
+    table_list = Metadata.bigquery.get_table_list(dto.connection_info)
+    return {"tables": table_list}
+
+
+@router.post("/metadata/constraints")
+@log_dto
+def get_bigquery_constraints(dto: MetadataDTO) -> dict:
+    table_list = Metadata.bigquery.get_constraints(dto.connection_info)
+    return {"constraints": table_list}
```
Comment (grieve54706) [inline ibis-server/app/routers/ibis/bigquery.py:39]: It doesn't have any other data. How about just returning a list to a JSON array?
BTW, plz add a empty line in the end of file.

**Comment by grieve54706** [comment]
I saw you use POST for API in the codebase, but your description of PR is not mapped.

File: ibis-server/tests/routers/ibis/test_bigquery.py [hunk 19]
```diff
@@ -220,3 +220,19 @@ def test_query_with_dry_run_and_invalid_sql(self):
...
+        )
+        assert response.status_code == 200
+
+    def test_metadata_list_constraints(self):
+        connection_info = self.get_connection_info()
+        response = client.post(
+            url="/v2/ibis/bigquery/metadata/constraints",
+            json={"connectionInfo": connection_info},
+        )
+        assert response.status_code == 200
```
Comment (grieve54706) [inline ibis-server/tests/routers/ibis/test_bigquery.py:238]: Please assert the contents.

Comment (onlyjackfrost) [reply ibis-server/tests/routers/ibis/test_bigquery.py:238]: I think we do not need to assert the content here, cause Pydentic will raise errors if the responding data structure is incorrect, and we do not care about the actual data in the data structure

Comment (grieve54706) [reply ibis-server/tests/routers/ibis/test_bigquery.py:238]: No, FastAPI only checks the response type is dict. We should check the content like the data count or fields like `name` and `column` should be with it.

Comment (onlyjackfrost) [reply ibis-server/tests/routers/ibis/test_bigquery.py:238]: I'll update the test case and modify the dataset content in another PR

**Review by grieve54706 — APPROVED** [review]

**Pull request merged**